kubernetes/deployments_kubeclient.go

```diff
@@ -0,0 +1,1339 @@
+		if !ok {
+			return errors.New("Spec missing from route")
+		}
+		// Determine which service this route points to
+		to, ok := spec["to"].(map[interface{}]interface{})
+		if !ok {
+			return errors.New("Route has no destination")
+		}
+		toName, ok := to["name"].(string)
+		if !ok || len(toName) == 0 {
+			return errors.New("Service name missing or invalid for route")
+		}
+
+		var matchingServices []string
+		// Check if this route is for a service we're interested in
+		_, pres := routesByService[toName]
+		if pres {
+			matchingServices = append(matchingServices, toName)
+		}
+
+		// Also check alternate backends for services
+		altBackends, ok := spec["alternateBackends"].([]interface{})
+		if ok {
+			for idx := range altBackends {
+				backend, ok := altBackends[idx].(map[interface{}]interface{})
+				if !ok {
+					return errors.New("Malformed alternative backend")
+				}
+				// Check if this alternate backend is a service we want a route for
+				backendKind, _ := backend["kind"].(string)
+				if backendKind == "Service" {
+					backendName, ok := backend["name"].(string)
+					if ok && len(backendName) > 0 {
+						_, pres := routesByService[backendName]
+						if pres {
+							matchingServices = append(matchingServices, backendName)
+						}
+					}
+				}
+			}
+		}
+		if len(matchingServices) > 0 {
+			// Get ingress points
+			status, ok := routeItem["status"].(map[interface{}]interface{})
+			if !ok {
+				return errors.New("Status missing from route")
+			}
+			ingresses, ok := status["ingress"].([]interface{})
+			if !ok {
+				return errors.New("No ingress array listed in route")
+			}
+
+			// Prefer ingress with oldest lastTransitionTime that is marked as admitted
+			oldestAdmittedIngress, err := findOldestAdmittedIngress(ingresses)
+			if err != nil {
+				return err
+			}
+
+			// Use hostname from oldest admitted ingress if possible
+			var hostname string
+			if oldestAdmittedIngress != nil {
+				hostname, ok = oldestAdmittedIngress["host"].(string)
+				if !ok {
+					return errors.New("Hostname missing from ingress")
+				}
+			} else {
+				// Fall back to optional host in spec
+				hostname, _ = spec["host"].(string)
+			}
+
+			// Check for optional path
+			path, _ := spec["path"].(string)
+
+			// Determine whether route uses TLS
+			// see: https://github.com/openshift/origin-web-console/blob/v3.7.0/app/scripts/filters/resources.js#L193
+			isTLS := false
+			tls, ok := spec["tls"].(map[interface{}]interface{})
+			if ok {
+				tlsTerm, ok := tls["termination"].(string)
+				if ok && len(tlsTerm) > 0 {
+					isTLS = true
+				}
+			}
+
+			// Check if this route uses a custom hostname
+			customHost := true
+			metadata, ok := routeItem["metadata"].(map[interface{}]interface{})
+			if ok {
+				annotations, ok := metadata["annotations"].(map[interface{}]interface{})
+				if ok {
+					hostGenerated, _ := annotations["openshift.io/host.generated"].(string)
```

Comment:
are you sure you can skip the type validation ?

Reply:
I'll change this to use two steps, first to check if the key is present and second to perform the type assertion.

Reply:
thanks